blog/2024-06-11-automap.md

```diff
@@ -0,0 +1,567 @@
+      M_2 <= b_M_2 * BIG_NUMBER
+      R_2 <= b_R_2 * BIG_NUMBER
+      
+Now we have ourselves a real ILP! To solve it, we can just feed it into any
+existing solver (there are *many*) or implement one of the many known algorithms
+for ILP solving. Technically, ILP is in NP (i.e., it has exponential runtime)
+but in practice solvers are very fast and the ILPs we generate are small, so
+even if you're solving a bunch of these ILPs when type checking a large program
+it still makes for a completely usable system.
+
+# Putting it all together
+
+We now have all the ingredients we need to bestow Futhark with the power of rank
+polymorphism. Here's how the rank polymorphism system works from
+start-to-finish:
+
+1. Traverse the code: for each application generate a rank equality constraint
+   and a Rule 1 constraint and add it to the constraint set.
+   
+2. Transform the constraint set into an ILP and solve it. If the ILP has no
+   solutions or multiple solutions, fail and report an error to the user.
+
+3. The solution to the ILP tells the compiler how to insert the minimal number
+   of `map`s and `rep`s to make the program rank-correct. For example, if `M_i =
+   3` and `R_i = 0` for an application `f x`, then transform the application by
+   inserting three `map`s and no `rep`s: `f x --> map (map (map f)) x`.
+   
+4. Type check the transformed program with all implicit `map`s and `rep`s made
+   explicit as before and continue with compilation.
+   
+The extra nice thing here is that we can do all this without affecting how the
+rest of the compiler works. This is a direct result of the fact that Futhark
+treats rank polymorphism as syntactic sugar and there's no "magic" going
+on---all implicit constructs (`map` and `rep`) are just standard functions
+available to the programmer regardless. This also means that the programmer
+is free to use the rank polymorphism system to whatever extent that they
+wish---from omitting all `map`s and `rep`s (up to ambiguity), to only some,
+to having a program with all `map`s and `rep`s explicit.
+
+We call this system AUTOMAP (implicit `map`s are way more common than implicit
+`rep`s and AUTOMAPREP doesn't sound as cool) and it should be coming to Futhark
+soon(ish)!
+   
+**Footnotes**
+
+
+[1] Of course you could be rigorous here and also add something along the lines of
+
+  We define `(x + y)_i = x_i + y_i` (where `x_i` is the `i`th component of `x`)
+
+so that the operation is well-defined. With this approach there are really two
+definitions of `+`---one for scalars and one for vectors and we've just
+grouped them under the same name.  From a programming languages perspective,
+this corresponds to `ad-hoc polymorphism
+<https://en.wikipedia.org/wiki/Ad_hoc_polymorphism>`_ (a.k.a function/operator
+overloading) wherein you group a bunch of functions (which expect different
+types) under the same name.
+
+[2] In some languages (like [Typed Remora](https://arxiv.org/abs/1912.13451)),
```

Comment:
I think that is the case in pretty much *every* language except for Futhark, and it is certainly the case in APL and the like where the term originates.